Add "Clash 1.2.2 and future plans" (#23)
clash-lang/clash-lang.org: content/news/04-clash122.md

---
title: "Clash 1.2.2 and future plans"
date: "2020-06-14"
description: "Bugs bugs, and future plans"
disable_comments: false
author: "martijnbastiaan"
authorbox: true # Optional, enable authorbox for specific post
summary: Last Friday we released Clash 1.2.2 on GitHub. In big part thanks to our lovely community, we've found and fixed many bugs. Thanks to some patches, Clash even runs anywhere from 2-20% faster for some common design patterns! With this blogpost I'd like to highlight some of the progress we've made since the 1.2 release, and elaborate on our future plans.
toc: false
mathjax: false
---

Hi everyone!

> Last Friday we released Clash 1.2.2 on GitHub. In big part thanks to our lovely community, we've found and fixed many bugs. Thanks to some patches, Clash even runs anywhere from 2-20% faster for some common design patterns! Of course, the full list of changes can be found in [the CHANGELOG](https://github.com/clash-lang/clash-compiler/releases/tag/v1.2.2) over at GitHub. With this blogpost I'd like to highlight some of the progress we've made since the 1.2 release, and elaborate on our future plans. 

Reviewer: ```suggestion
Last Friday we cut version 1.2.2 of Clash on [GitHub](https://github.com/clash-lang/clash-compiler/releases/tag/v1.2.2), uploaded a source release on [Hackage](http://hackage.haskell.org/package/clash-ghc-1.2.2) and a binary release on [Snapcraft](https://snapcraft.io/clash). In big part thanks to our lovely community, we've found and fixed many bugs. Thanks to some patches, Clash even runs anywhere from 2-20% faster for some common design patterns! Of course, the full list of changes can be found in [the CHANGELOG](https://github.com/clash-lang/clash-compiler/releases/tag/v1.2.2) over at GitHub. With this blogpost we'd like to highlight some of the progress we've made since the 1.2 release, and elaborate on our future plans. 
```

Reviewer 2: "we cut version 1.2.2 on GitHub" -> "we _put_"

I can't propose a change to a proposed change by the looks of it :-)

# Starter project
With the release of Clash 1.2.2 we've also released the first iteration of our [starter project](https://github.com/clash-lang/clash-compiler/blob/b9cd32cd876eaee1d5591f40c20b1238be195159/clash-starters/clash-example-project/README.md)! The goal of this starter project is to provide a fresh template for a new Clash project. Dependencies can now be added exactly like you would for a Haskell project; simply by adding it to your dependency list in the project's cabal file. Whereas previously you'd have to write custom scripts, or rely on nix to solve it for you!

# Memory leak begone

> For a while now we've seen pretty bad memory leaks in commercial deployments of Clash, but haven't been able to replicate this in a small example. Through various workarounds we've been able to reduce the memory usage to make a usable on "reasonable" hardware and moved on to other, by that time higher priority, issues. We were recently confronted with this issue again, reported on our [public issue tracker](https://github.com/clash-lang/clash-compiler/issues/1256). This time it even had a small example! Time to investigate.

Reviewer 2: ```suggestion
For a while now we've seen pretty bad memory leaks in commercial deployments of Clash, but haven't been able to replicate this in a small example. Through various workarounds we've been able to reduce the memory usage to make it usable on "reasonable" hardware and moved on to other, by that time higher priority, issues. We were recently confronted with this issue again, reported on our [public issue tracker](https://github.com/clash-lang/clash-compiler/issues/1256). This time it even had a small example! Time to investigate.
```

Reviewer 2: Suggested change: "to make a usable on [...]" -> "to make it usable on [...]"

> For a while now we've seen pretty bad memory leaks in commercial deployments of Clash, but haven't been able to replicate this in a small example. Through various workarounds we've been able to reduce the memory usage to make a usable on "reasonable" hardware and moved on to other, by that time higher priority, issues. We were recently confronted with this issue again, reported on our [public issue tracker](https://github.com/clash-lang/clash-compiler/issues/1256). This time it even had a small example! Time to investigate.

Reviewer: ```suggestion
For a while now we've seen pretty bad memory leaks during simulation in commercial deployments of Clash, but haven't been able to replicate this in a small example. Through various workarounds we've been able to reduce the memory usage to make a usable on "reasonable" hardware and moved on to other, by that time higher priority, issues. We were recently confronted with this issue again, reported on our [public issue tracker](https://github.com/clash-lang/clash-compiler/issues/1256). This time it even had a small example! Time to investigate.
```

> Clash designs operate on infinite streams of data called [signals](https://hackage.haskell.org/package/clash-prelude-1.2.2/docs/Clash-Signal.html). Signals model (bundles of) wires, where each value corresponds to a stabilized signal at the end of a clock cycle. Signals do not always carry known values. For example, in some hardware configurations the very first value produced by a memory element is unknown. For this purpose, Clash exports a bottom called [XException](https://hackage.haskell.org/package/clash-prelude-1.2.2/docs/Clash-XException.html). This presents a problem: for performance reasons you might want to evaluate a value to _Weak Head Normal Form_ (WHNF) or even _Normal Form_ (NF). Usually, you'd use `seq` or `deepseq` to do this, but for typical values used in Clash you can't, as evaluating an `XException` to (WH)NF would stop your circuit's simulation dead in its tracks! To work around this, Clash 1.0 [introduced](/blog/0004-undefined-values/) a type class [NFDataX](https://hackage.haskell.org/package/clash-prelude-1.2.2/docs/Clash-XException.html#t:NFDataX). It allows us to variants of `seq` and `deepseq` that simply ignore `XExceptions`, but otherwise do what you except them to do. Where you'd normally write:

Reviewer: ```suggestion
Clash designs operate on infinite streams of data called [signals](https://hackage.haskell.org/package/clash-prelude-1.2.2/docs/Clash-Signal.html). Signals model (bundles of) wires, where each value corresponds to a stabilized signal at the end of a clock cycle. Signals do not always carry known values. For example, in some hardware configurations the very first value produced by a memory element is unknown. For this purpose, Clash exports an exception-throwing "unknown" value called [XException](https://hackage.haskell.org/package/clash-prelude-1.2.2/docs/Clash-XException.html). This presents a problem: for performance reasons you might want to evaluate a value to _Weak Head Normal Form_ (WHNF) or even _Normal Form_ (NF). Usually, you'd use `seq` or `deepseq` to do this, but for typical values used in Clash you can't, as evaluating an `XException` to (WH)NF would stop your circuit's simulation dead in its tracks! To work around this, Clash 1.0 [introduced](/blog/0004-undefined-values/) a type class [NFDataX](https://hackage.haskell.org/package/clash-prelude-1.2.2/docs/Clash-XException.html#t:NFDataX). It allows us to variants of `seq` and `deepseq` that simply ignore `XExceptions`, but otherwise do what you expect them to do. Where you'd normally write:
```

> Clash designs operate on infinite streams of data called [signals](https://hackage.haskell.org/package/clash-prelude-1.2.2/docs/Clash-Signal.html). Signals model (bundles of) wires, where each value corresponds to a stabilized signal at the end of a clock cycle. Signals do not always carry known values. For example, in some hardware configurations the very first value produced by a memory element is unknown. For this purpose, Clash exports a bottom called [XException](https://hackage.haskell.org/package/clash-prelude-1.2.2/docs/Clash-XException.html). This presents a problem: for performance reasons you might want to evaluate a value to _Weak Head Normal Form_ (WHNF) or even _Normal Form_ (NF). Usually, you'd use `seq` or `deepseq` to do this, but for typical values used in Clash you can't, as evaluating an `XException` to (WH)NF would stop your circuit's simulation dead in its tracks! To work around this, Clash 1.0 [introduced](/blog/0004-undefined-values/) a type class [NFDataX](https://hackage.haskell.org/package/clash-prelude-1.2.2/docs/Clash-XException.html#t:NFDataX). It allows us to variants of `seq` and `deepseq` that simply ignore `XExceptions`, but otherwise do what you except them to do. Where you'd normally write:

Reviewer 2: ```suggestion
Clash designs operate on infinite streams of data called [signals](https://hackage.haskell.org/package/clash-prelude-1.2.2/docs/Clash-Signal.html). Signals model (bundles of) wires, where each value corresponds to a stabilized signal at the end of a clock cycle. Signals do not always carry known values. For example, in some hardware configurations the very first value produced by a memory element is unknown. For this purpose, Clash exports a bottom called [XException](https://hackage.haskell.org/package/clash-prelude-1.2.2/docs/Clash-XException.html). This presents a problem: for performance reasons you might want to evaluate a value to _Weak Head Normal Form_ (WHNF) or even _Normal Form_ (NF). Usually, you'd use `seq` or `deepseq` to do this, but for typical values used in Clash you can't, as evaluating an `XException` to (WH)NF would stop your circuit's simulation dead in its tracks! To work around this, Clash 1.0 [introduced](/blog/0004-undefined-values/) a type class [NFDataX](https://hackage.haskell.org/package/clash-prelude-1.2.2/docs/Clash-XException.html#t:NFDataX). It allows us to use variants of `seq` and `deepseq` that simply ignore `XExceptions`, but otherwise do what you except them to do. Where you'd normally write:
```

Reviewer 2: "allows us to variants" -> "allows us to use variants"

```
a `seq` b
```

> you'd now be able to write:

Reviewer: ```suggestion
you'd write the following instead:
```

```
a `defaultSeqX` a :- as
```

The purpose of this expression is to force a (WH)NF evaluation of `a` before being able to observe the _spine_ of the signal. I.e., evaluate a value at time _t_ if you want to observe the value at _t+1_. This solves the same problem as the one described under [building up unevaluated expressions](https://wiki.haskell.org/Memory_leak) at the Haskell wiki. It turns out that this expression was parsed as:

```
(a `defaultSeqX` a) :- as
```

instead of what we actually wanted and expected:

```
a `defaultSeqX` (a :- as)
```

We've since added ```infixr 0 `defaultSeqX` ```. The expression now parses as expected and we've observed that this fixes the memory issues our clients were seeing! It turns out that `deepseq` has exactly the same problem, for which we've [submitted a patch](https://github.com/haskell/deepseq/pull/56). 

In short, huge thanks to [Dr. Gergő Érdi](https://github.com/gergoerdi) who pushed us to get to the bottom of the issue and release a fix for it 🎉.

# Plans for 1.4
The next major relase of Clash will be 1.4. Two major additions are in the pipeline:

> 1. _Ground work for a Partial Evaluator._ Clash currently uses a transformation based rewrite system to translate Haskell into synthesizable circuits. This transformation system was [originally written](https://research.utwente.nl/en/publications/digital-circuit-in-c%CE%BBash-functional-specifications-and-type-direc) to remove polymorphism from GHC's Core output. Bit by bit, more and more transformations were added: removing case statements based on a known subject, unrolling primitive definitions, removing lambdas by propagating their arguments, and many more. As the number of transformations grew, so did the complexity and predictability. In fact, we're currently seeing issues that are very hard to solve within a transformation based system.

Reviewer 2: ```suggestion
1. _Ground work for a Partial Evaluator._ Clash currently uses a transformation based rewrite system to translate Haskell into synthesizable circuits. This transformation system was [originally written](https://research.utwente.nl/en/publications/digital-circuit-in-c%CE%BBash-functional-specifications-and-type-direc) to remove polymorphism from GHC's Core output. Bit by bit, more and more transformations were added: removing case statements based on a known subject, unrolling primitive definitions, removing lambdas by propagating their arguments, and many more. As the number of transformations grew, so did the complexity and unpredictability. In fact, we're currently seeing issues that are very hard to solve within a transformation based system.
```

Reviewer 2: "complexity and predictability grew" -> "... unpredictability"

> 1. _Ground work for a Partial Evaluator._ Clash currently uses a transformation based rewrite system to translate Haskell into synthesizable circuits. This transformation system was [originally written](https://research.utwente.nl/en/publications/digital-circuit-in-c%CE%BBash-functional-specifications-and-type-direc) to remove polymorphism from GHC's Core output. Bit by bit, more and more transformations were added: removing case statements based on a known subject, unrolling primitive definitions, removing lambdas by propagating their arguments, and many more. As the number of transformations grew, so did the complexity and predictability. In fact, we're currently seeing issues that are very hard to solve within a transformation based system.

Reviewer: ```suggestion
1. _Ground work for a Partial Evaluator._ Clash currently uses a transformation based rewrite system to translate Haskell into synthesizable circuits. This transformation system was [originally written](https://research.utwente.nl/en/publications/digital-circuit-in-c%CE%BBash-functional-specifications-and-type-direc) to make all terms in GHC's/Clash's Core IR completely monomorphic and first-order. Bit by bit, more and more transformations were added: removing case statements based on a known subject, unrolling primitive definitions, removing lambdas by propagating their arguments, and many more. As the number of transformations grew, so did the complexity and predictability. In fact, we're currently seeing issues that are very hard to solve within a transformation based system.
```

> 1. _Ground work for a Partial Evaluator._ Clash currently uses a transformation based rewrite system to translate Haskell into synthesizable circuits. This transformation system was [originally written](https://research.utwente.nl/en/publications/digital-circuit-in-c%CE%BBash-functional-specifications-and-type-direc) to make all terms in GHC's/Clash's Core IR completely monomorphic and first-order. Bit by bit, more and more transformations were added: removing case statements based on a known subject, unrolling primitive definitions, removing lambdas by propagating their arguments, and many more. As the number of transformations grew, so did the complexity and predictability. In fact, we're currently seeing issues that are very hard to solve within a transformation based system.

Reviewer 3: > As the number of transformations grew, so did the complexity and predictability.
You meant the predictability is going down?

Reviewer 2: It appears two of my copyedit suggestions got lost somewhere in the commits. I'll re-add them now.

Reviewer 4: It's predictably less predictable.

Author: Fixed!

> 1. _Ground work for a Partial Evaluator._ Clash currently uses a transformation based rewrite system to translate Haskell into synthesizable circuits. This transformation system was [originally written](https://research.utwente.nl/en/publications/digital-circuit-in-c%CE%BBash-functional-specifications-and-type-direc) to make all terms in GHC's/Clash's Core IR completely monomorphic and first-order. Bit by bit, more and more transformations were added: removing case statements based on a known subject, unrolling primitive definitions, removing lambdas by propagating their arguments, and many more. As the number of transformations grew, so did the complexity and predictability. In fact, we're currently seeing issues that are very hard to solve within a transformation based system.

Reviewer 2: ```suggestion
1. _Ground work for a Partial Evaluator._ Clash currently uses a transformation based rewrite system to translate Haskell into synthesizable circuits. This transformation system was [originally written](https://research.utwente.nl/en/publications/digital-circuit-in-c%CE%BBash-functional-specifications-and-type-direc) to make all terms in GHC's/Clash's Core IR completely monomorphic and first-order. Bit by bit, more and more transformations were added: removing case statements based on a known subject, unrolling primitive definitions, removing lambdas by propagating their arguments, and many more. As the number of transformations grew, so did the complexity and unpredictability. In fact, we're currently seeing issues that are very hard to solve within a transformation based system.
```

Reviewer 2: It appears this copyedit suggestion I made got lost somewhere in the heap of small changes :-). I suggested this change: "complexity and predictability grew" -> unpredictability

Author: oh oops, sorry

Reviewer 2: I don't think the GitHub review system works very well on Markdown... it's more for < 80 columns of code, rather than full paragraphs... so I blame the system :-)

Author: Fixed!

>    Luckily, many of these transformations are basically trying to do one thing: evaluate a Haskell core expression. By building a partial evaluator, we hope to greatly reduce the complexity of Clash. We believe this should, when fully implemented, improve synthesis times by at least an order of magnitude. We don't expect it to be fully implemented before Clash 1.8 however.

Reviewer: ```suggestion
   Luckily, many of these transformations are basically trying to do one thing: evaluate a Haskell core expression. By building a partial evaluator, we hope to greatly reduce the complexity of Clash. We believe this should, when fully implemented, improve synthesis times by at least an order of magnitude. We don't expect it to be fully implemented before Clash 1.6 however.
```
